src/concurrency.md

```diff
@@ -232,6 +233,96 @@ fn run() -> Result<()> {
 # quick_main!(run);
 ```
 
+[threadpool-walk]: #threadpool-walk
+<a name="threadpool-walk"></a>
+## Use a threadpool to calculate SHA1 to all *.iso files in a folder
+
+[![threadpool-badge]][threadpool] [![num_cpus-badge]][num_cpus] [![walkdir-badge]][walkdir] [![cat-concurrency-badge]][cat-concurrency][![cat-filesystem-badge]][cat-filesystem]
+
+This example uses `threadpool`, `walkdir` and `ring` crates to calculate the SHA1 for every file present in the current directory. A threadpool `pool` is created using the number of cpus present in the system. Then every `entry` returned by `walkdir` is passed into this pool to perform the operations of reading and compute SHA1. At the end the program waits for all jobs to finish. To get better results, compile this program in release mode.
+
+```rust,no_run
+# #[macro_use]
+# extern crate error_chain;
+extern crate walkdir;
+extern crate ring;
+extern crate num_cpus;
+extern crate threadpool;
+
+# error_chain! {
+#     foreign_links {
+#         Io(std::io::Error);
+#     }
+# }
+
+use walkdir::{WalkDir, DirEntry};
+use std::fs::File;
+use std::io::BufReader;
+use std::io::prelude::*;
+use threadpool::ThreadPool;
+use std::sync::mpsc::channel;
+use ring::digest;
+
+# // Verify the iso extension
+# fn is_iso(entry: &DirEntry) -> bool {
+#     entry.file_name()
+#         .to_str()
+#         .map(|s| s.ends_with(".iso"))
+#         .unwrap_or(false)
+# }
+
+fn run() -> Result<()> {
+    let cpus = num_cpus::get();
+    let pool = ThreadPool::new(cpus);
+    let mut file_counter: usize = 0;
+
+    let (tx, rx) = channel();
+
+    // Look in the current directory.
+    for entry in WalkDir::new(".")
+        .follow_links(true)
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .filter(|e| !e.file_type().is_dir() && is_iso(e)) {
```

Comment:
- I would operate on `Path`or `PathBuf` instead of `DirEntry` here It is lighter, more idiomatic and has all the required interfaces to boot :).
- I would split this closure to `filter` (only the condition) and `map` (the actual body). It would convey the intent better.